lib/core/src/display/symbol/shader.rs

```diff
@@ -0,0 +1,181 @@
+
+    /// Creates new shader with attached callback.
+    pub fn new(logger:Logger, stats:&Stats, context:&Context, on_mut:OnMut) -> Self {
+        stats.inc_shader_count();
+        let geometry_material = default();
+        let material          = default();
+        let program           = default();
+        let dirty_logger      = logger.sub("dirty");
+        let dirty             = Dirty::new(dirty_logger,on_mut);
+        let context           = context.clone();
+        let stats             = stats.clone_ref();
+        dirty.set();
+        Self {geometry_material,material,program,dirty,logger,context,stats}
+    }
+
+    /// Check dirty flags and update the state accordingly.
+    pub fn update(&mut self, bindings:&[VarBinding]) {
+        group!(self.logger, "Updating.", {
+            if self.dirty.check_all() {
+
+                self.stats.inc_shader_compile_count();
+
+                let mut shader_cfg     = shader::builder::ShaderConfig::new();
+                let mut shader_builder = shader::builder::ShaderBuilder::new();
+
+                for binding in bindings {
+                    let name = &binding.name;
+                    let tp   = &binding.decl.tp;
+                    match binding.scope {
+                        None => todo!(),
+                        Some(scope_type) => match scope_type {
+                            ScopeType::Instance => shader_cfg.add_attribute (name,tp),
+                            ScopeType::Point    => shader_cfg.add_attribute (name,tp),
+                            ScopeType::Global   => shader_cfg.add_uniform   (name,tp),
+                            _ => todo!()
+                        }
+                    }
+                }
+
+                self.geometry_material.outputs().iter().for_each(|(name,decl)|{
+                    shader_cfg.add_shared_attribute(name,&decl.tp);
+                });
+
+                shader_cfg.add_output("color", glsl::PrimType::Vec4);
+
+                let vertex_code   = self.geometry_material.code().clone();
+                let fragment_code = self.material.code().clone();
+                shader_builder.compute(&shader_cfg,vertex_code,fragment_code);
+                let shader      = shader_builder.build();
+                let vert_shader = webgl::compile_vertex_shader  (&self.context,&shader.vertex);
+                let frag_shader = webgl::compile_fragment_shader(&self.context,&shader.fragment);
+                if let Err(ref err) = frag_shader {
+                    self.logger.error(|| format!("{}", err))
+                }
+
+                let vert_shader = vert_shader.unwrap();
+                let frag_shader = frag_shader.unwrap();
+                let program     = webgl::link_program(&self.context,&vert_shader,&frag_shader);
+
+                let program     = program.unwrap();
+                self.program    = Some(program);
+                self.dirty.unset_all();
+            }
+        })
```

Comment:
Another very long function

Reply:
Good catch. Very WIP function. Addded note that it should be splitted when finished in next PR